### PR TITLE
feat(sdk, api, ui): deployment platform

### PR DESCRIPTION
**Pull request by fsamin**
1. Description
Ability to run "deploy application" on a platform... and this is magic
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by fsamin** [comment]
TODO:
* Implement provider token mechanism
* fix unit tests
* implement a platform plugin for our first provider (see sdk/grpcplugin/platformplugin/example)
* fix delete application deployment strategy handler
* fix many subtle things in UI (translation, toaster, cache, etc.)

**Comment by bnjjj** [comment]
@fsamin let me know what to do in UI I will do that :)

**Comment by fsamin** [comment]
Thanks !

Try to edit deployment strategies on an application (application/advanded) and please don't blame me.
To be able to do it you have to:
* Import a deployment provider
* Attach the deployment platform to a project (project/platform)

To import a deployment provider, create a yml file with:
```yaml
name: my-platform
default_config:
  url: 
    type: string
    value: "https://localhost:12345"
  user: 
    type: string
    value: "user"
  password: 
    type: password
deployment: true
deployment_default_config: 
  deployment_config: 
    type: string
  secret_token:
    type: password
plugin: test-plugin
```

then import it with the CLI:
```
cdsctl admin platform-model import my-pf.yml
```

After that, you can go to Project/Platform and add the provider to your project.
Then, on an application, you can set the deployment strategies in Application/Advanced

**Comment by fsamin** [comment]
TODO:
* ~~Implement provider token mechanism~~
* ~~fix unit tests~~
* implement a platform plugin for our first provider (see sdk/grpcplugin/platformplugin/example)
* fix delete application deployment strategy handler
* fix many subtle things in UI (translation, toaster, cache, etc.)
* Implement deployment strategies in "as code" feature

File: engine/api/application_deployment.go [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+			return sdk.WrapError(err, "postApplicationDeploymentStrategyConfigHandler> unable to load application")
+		}
+
+		if err := tx.Commit(); err != nil {
+			return sdk.WrapError(err, "postApplicationDeploymentStrategyConfigHandler> unable to commit tx")
+		}
+
+		if getProvider(ctx) != nil {
+			p := getProvider(ctx)
+			log.Info("postApplicationDeploymentStrategyConfigHandler> application %s configuration successfully ypdated by provider %s", appName, *p)
```
Comment (yesnault) [inline engine/api/application_deployment.go]: ypdated -> updated

File: sdk/plugin.go [hunk 2]
```diff
@@ -101,3 +101,51 @@ func DeletePlugin(name string) error {
...
+	GRPCPluginAction             = "action"
+)
+
+type GRPCPlugin struct {
+	ID          int64              `json:"id" cli:"id" db:"id"`
+	Name        string             `json:"name" cli:"name,key" db:"name"`
+	Type        string             `json:"type" cli:"type" db:"type"`
+	Author      string             `json:"author" cli:"author" db:"author"`
+	Description string             `json:"description" cli:"description" db:"description"`
+	Binaries    []GRPCPluginBinary `json:"binaries" cli:"_" db:"-"`
```
Comment (yesnault) [inline sdk/plugin.go]: underscore for cli?

**Comment by fsamin** [comment]
TODO:
* ~~Implement provider token mechanism~~
* ~~fix unit tests~~
* implement a platform plugin for our first provider (see sdk/grpcplugin/platformplugin/example)
* ~~fix delete application deployment strategy handler~~
* fix many subtle things in UI (translation, toaster, cache, etc.)
* Implement deployment strategies in "as code" feature

**Comment by fsamin** [comment]
TODO:
* ~~Implement provider token mechanism~~
* ~~fix unit tests~~
* implement a platform plugin for our first provider (see sdk/grpcplugin/platformplugin/example)
* ~~fix delete application deployment strategy handler~~
* fix many subtle things in UI (translation, toaster, cache, etc.)
* ~~Implement deployment strategies in "as code" feature~~

File: engine/sql/091_deployment_platform.sql [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+);
+
+select create_unique_index('grpc_plugin', 'IDX_GRPC_PLUGIN_NAME', 'name');
+
+SELECT create_foreign_key('FK_PLATFORM_MODEL_GRPC_PLUGIN', 'platform_model', 'grpc_plugin', 'grpc_plugin_id', 'id');
+
+ALTER TABLE workflow_node_run_job ADD COLUMN platform_plugin_binaries JSONB;
+
+-- +migrate Down
+ALTER TABLE worker_model DROP COLUMN grpc_plugin_id;
```
Comment (yesnault) [inline engine/sql/091_deployment_platform.sql]: it's platform_model, not worker_model in this alter table

File: engine/sql/091_deployment_platform.sql [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ALTER TABLE worker_model DROP COLUMN deployment_default_config;
```
Comment (yesnault) [inline engine/sql/091_deployment_platform.sql]: it's platform_model, not worker_model in this alter table

File: engine/sql/091_deployment_platform.sql [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+DROP TABLE application_deployment_strategy;
+ALTER TABLE workflow_node_context DROP COLUMN project_platform_id;
+DROP TABLE grpc_plugin;
```
Comment (yesnault) [inline engine/sql/091_deployment_platform.sql]: missing drop column platform_plugin_binaries from workflow_node_run_job in rollback

File: ui/src/assets/i18n/fr.json [hunk 6]
```diff
@@ -88,6 +88,7 @@
   "application_repoman_cant_link": "ici",
   "application_repo_title": "Lier l'application à un repository : ",
   "application_repo_strat_title" : "Configuration",
+  "application_deployment_title" : "Lier l'application à une plateforme de déploiement: ",
```
Comment (yesnault) [inline ui/src/assets/i18n/fr.json]: key deployment_platform_name is missing (used in application -> advanced)

File: ui/src/assets/i18n/en.json [hunk 7]
```diff
@@ -88,6 +88,7 @@
   "application_repoman_cant_link": "here",
   "application_repo_title": "Link application to a repository: ",
   "application_repo_strat_title" : "Configuration",
+  "application_deployment_title" : "Link application to a deployment platform: ",
```
Comment (yesnault) [inline ui/src/assets/i18n/en.json]: key deployment_platform_name is missing (used in application -> advanced)

File: ui/src/assets/i18n/fr.json (hunk 6 above)
Comment (fsamin) [reply ui/src/assets/i18n/fr.json]: done

File: ui/src/assets/i18n/en.json (hunk 7 above)
Comment (fsamin) [reply ui/src/assets/i18n/en.json]: done

**Comment by fsamin** [comment]
TODO:
* ~~Implement provider token mechanism~~
* ~~fix unit tests~~
* implement a platform plugin for our first provider (see sdk/grpcplugin/platformplugin/example)
* ~~fix delete application deployment strategy handler~~
* ~~fix many subtle things in UI (translation, toaster, cache, etc.)~~
* ~~Implement deployment strategies in "as code" feature~~

File: engine/api/application/application_platform.go [hunk 8]
```diff
@@ -69,6 +69,30 @@ func LoadDeploymentStrategies(db gorp.SqlExecutor, appID int64, withClearPasswor
...
+	query := `DELETE FROM application_deployment_strategy 
+	WHERE application_id = $1
+	AND project_platform_id IN (
+		SELECT 	project_platform.id 
+		FROM project_platform
+		WHERE project_platform.project_id = $2
+		AND project_platform.platform_model_id = $3
+	)`
+
+	if _, err := db.Exec(query, appID, projID, pfID); err != nil {
```
Comment (bnjjj) [inline engine/api/application/application_platform.go]: useless if

File: engine/api/workflow/execute_node_job_run.go [hunk 9]
```diff
@@ -313,6 +314,49 @@ func LoadNodeJobRunSecrets(db gorp.SqlExecutor, store cache.Store, job *sdk.Work
...
+						Value: v.Value,
+					})
+				}
+			}
+			apv = sdk.VariablesPrefix(apv, "cds.platform.")
+			apv = sdk.VariablesFilter(apv, sdk.SecretVariable)
+		}
+
+	}
+	secrets = append(secrets, pfv...)
```
[the remark this answers is not in the document]
Comment (fsamin) [reply engine/api/workflow/execute_node_job_run.go]: #2664

**Pull request merged**